Backpropagate Cargo.lock updates to all lock files

ryoqun opened the pull request:
#### Problem

Non-default Cargo lock files tend to be forgot to update. this causes many problems:
  - CI build isn't deterministic because lock files are changed on the fly while ci is running; So, `git diff` on CI shows some difference.
  - auditing cumulative Cargo lock update is very tedious or hard: https://github.com/solana-labs/solana/commit/959c1ea857b7a52c590f7d02efaa09eb041e732c#diff-7d7ecaa361c7585f6ee9b01378cd1393R2872
  - back-porting sometimes incurs noise of unrelated previous cargo lock updates, which is again tedious to audit: https://github.com/solana-labs/solana/commit/d2107270eaae80ac44cdf3998564e58a49db2654#diff-d2ede298d9a75c849170d6ef285eda1eR6578
- dependabot sometimes fails because it can't update non-default Cargo lock files

This is a problem which I noticed at https://github.com/solana-labs/solana/pull/8882. Since then, I've been always bothered with our trust-chain in the build/CI system a bit... Let's fortify it step by step before we get pray of third-party crate publisher's credential hack :)

#### Summary of Changes
- never accept stale lock files
 - for that, give special handling to dependabot

live example: https://github.com/solana-labs/solana/pull/9159/commits


partly related to this: https://github.com/solana-labs/solana/issues/8587#issuecomment-597376904

garious commented on ci/docker-run.sh:
```diff
@@ -76,6 +76,19 @@ ARGS+=(
+if [[ $BUILDKITE_LABEL == "checks" ]] && echo "$BUILDKITE_BRANCH" | grep -E '^pull/[0-9]+/head$'; then
+  api_base="https://api.github.com/repos/solana-labs/solana/pulls"
+  pr_num=$(echo "$BUILDKITE_BRANCH" | grep -Eo '[0-9]+')
+  branch=$(curl -s "$api_base/$pr_num" | ruby -r json -e 'puts JSON.parse(STDIN.read())["head"]["ref"]')
```
I don't recall a Ruby dependency anywhere else, so would prefer not to add it here. Can you do this with Python?

ryoqun replied:
sure thing! Moreover, I wanted to ask we can install `jq` someway? If not, I'll go to the `python` path. :)

garious replied:
As tiny as jq is, I'd prefer we stick to dependencies that are available by default on Ubuntu and MacOS. If that restriction isn't practical, add jq to the Dockerfile.

mvines replied:
I'm ok with requiring `jq` (but yes it would need to be added to `ci/docker-rust/Dockerfile`), but `python` would be just fine too.

ryoqun replied:
Ok, I just switched to `python`. :)

ryoqun commented on ci/docker-run.sh:
```diff
@@ -96,4 +109,28 @@ fi
+      GIT_COMMITTER_NAME="$NAME" \
+      EMAIL="dependabot-buildkite@noreply.solana.com" \
+      git commit -m "Update all Cargo lock files"
+    git push origin "HEAD:$branch"
```
To make it work, this is escaped from more proper place like `test-checks.sh` (maybe comment-in-source so)

ryoqun commented on ci/test-checks.sh:
```diff
@@ -15,6 +15,30 @@ _ git show HEAD --check --oneline
+      _ scripts/cargo-for-all-lock-files.sh update $parsed_update_args
+    fi
+  fi
+  if _ scripts/cargo-for-all-lock-files.sh check --locked; then
```
This is the main dish of this PR.

ryoqun commented on ci/docker-run.sh:
```diff
@@ -76,6 +76,19 @@ ARGS+=(
+  branch=$(curl -s "$api_base/$pr_num" | ruby -r json -e 'puts JSON.parse(STDIN.read())["head"]["ref"]')
+  source_repo=$(curl -s "$api_base/$pr_num" | ruby -r json -e 'puts JSON.parse(STDIN.read())["head"]["repo"]["full_name"]')
+
+  if [[ $source_repo == "solana-labs/solana" ]] && echo "$branch" | grep "^dependabot/cargo/"; then
```
As this `if` gate is very sensitive to security as this allows the `git push` codepath.

mvines commented:
This seems really complex :-/

Maybe unchecking this box will help?
![image](https://user-images.githubusercontent.com/1281082/78064938-6144ea00-7347-11ea-9765-48daa6c6c98d.png)

Also isn't there a new Cargo.lock format available now that we could enable that'll reduce the backport conflicts?

ryoqun commented:
> This seems really complex :-/

Yeah, but I researched this a bit and no better solution came up for weeks while pondering when waiting builds.

> Maybe unchecking this box will help?
> ![image](https://user-images.githubusercontent.com/1281082/78064938-6144ea00-7347-11ea-9765-48daa6c6c98d.png)

I think the `top-level` here means the crates specified directly by `Cargo.toml`. So, non-top-level packages is the so-called transitively-dependant packages. So, this isn't what we want. Unchecking that would produce tons of tons dependabot PRs because plain `cargo update` produces so large diff locally already. Also, that would increase frequency of PRs even after we got over the initial surge.

For the dependabot fiddling path, I could add each `Cargo.{toml,lock}`s but it'd cause separate PRs just for the same crate update until they [support grouping of update](https://github.com/dependabot/feedback/issues/5).


> Also isn't there a new Cargo.lock format available now that we could enable that'll reduce the backport conflicts?

Yeah, that's true. But that doesn't solve all problems I'd like to fix as described above. Can do as a separate PR for the format change.

mvines commented:
Ok so I think my main concern about this PR is that how it adds a ton of stuff into `ci/docker-run.sh` and `ci/test-checks.sh`, where instead I think ideally this is all very isolated in it's own `ci/dependabot-pr.sh`-like file.

I think the ideal end state is a new build step in https://github.com/solana-labs/solana/blob/master/ci/buildkite.yml, that:
a. runs `ci/dependabot-pr.sh` before checks, with a "wait" directive,
b. but only runs for dependabot PRs

To get there will be a multi-step process:
1. We aren't surfacing all the right build environment variables to buildkite from ci-gate in order to implement 1.    https://github.com/mvines/ci-gate/blob/51d99887ea7cf3dd1d030a55cc6980f1fc8d09c7/index.js#L121-L127 is missing the "author" information (cc: https://buildkite.com/docs/apis/rest-api/builds#create-a-build) in particular, so all builds appear to be coming from me 🙃.   
2. Buildkite has support for build step conditionals, https://buildkite.com/docs/pipelines/conditionals, that will allow us to filter out this special build step from non-dependabot PRs (`build.env()` in particular)

I can look at adding (1) if this approach sounds good to you.

ryoqun commented:
> Ok so I think my main concern about this PR is that how it adds a ton of stuff into `ci/docker-run.sh` and `ci/test-checks.sh`, where instead I think ideally this is all very isolated in it's own `ci/dependabot-pr.sh`-like file.
> 
> I think the ideal end state is a new build step in https://github.com/solana-labs/solana/blob/master/ci/buildkite.yml, that:
> a. runs `ci/dependabot-pr.sh` before checks, with a "wait" directive,
> b. but only runs for dependabot PRs
> 
> To get there will be a multi-step process:
> 
>     1. We aren't surfacing all the right build environment variables to buildkite from ci-gate in order to implement 1.    https://github.com/mvines/ci-gate/blob/51d99887ea7cf3dd1d030a55cc6980f1fc8d09c7/index.js#L121-L127 is missing the "author" information (cc: https://buildkite.com/docs/apis/rest-api/builds#create-a-build) in particular, so all builds appear to be coming from me upside_down_face.
> 
>     2. Buildkite has support for build step conditionals, https://buildkite.com/docs/pipelines/conditionals, that will allow us to filter out this special build step from non-dependabot PRs (`build.env()` in particular)
> 
> 
> I can look at adding (1) if this approach sounds good to you.

Yeah, that looks great way to accomplish what I want to achieve. :)

ryoqun commented:
@mvines I think this pr is ready for review again!

I'm playing with this dependabot's PR this time: https://github.com/solana-labs/solana/pull/9508/commits

test builds:

this pr (for normal users): https://buildkite.com/solana-labs/solana/builds/22845
test subject pr build take 1 (it backpropagates updates to all locks): https://buildkite.com/solana-labs/solana/builds/22842
test subject pr build take 2: https://buildkite.com/solana-labs/solana/builds/22843

ryoqun commented on ci/test-checks.sh:
```diff
@@ -13,6 +13,15 @@ export RUSTFLAGS="-D warnings"
 # Look for failed mergify.io backports
 _ git show HEAD --check --oneline
 
+if _ scripts/cargo-for-all-lock-files.sh check --locked; then
```
This is what I really want to do before any our depending crates are compromised. :)

mvines commented on scripts/cargo-for-all-lock-files.sh:
```diff
@@ -0,0 +1,18 @@
+  files="$(git ls-files :**/Cargo.lock)"
+fi
+
+for lock_file in $files
+do
```
```suggestion
for lock_file in $files; do
```

mvines commented on ci/dependabot-updater.sh:
```diff
@@ -0,0 +1,20 @@
+package=$(echo "$parsed_update_args" | awk '{print $2}')
+if [[ -n $parsed_update_args ]]; then
+  # shellcheck disable=SC2086
+  _TARGET_LOCK_FILES=$(git grep --files-with-matches "$package" :**/Cargo.lock) \
```
Using `_TARGET_LOCK_FILES` to pass a list into `scripts/cargo-for-all-lock-files.sh` feels a little too sneaky to me.  It would be nice to fit this on the `scripts/cargo-for-all-lock-files.sh` command line.

`$ scripts/cargo-for-all-lock-files.sh $(_TARGET_LOCK_FILES) -- update $parsed_update_args` perhaps?  And if there's no `--` argument then we use `$(git ls-files :**/Cargo.lock)`

ryoqun replied:
Yeah, I've been concerned a bit too... I followed an easier path considering these script's intended usage frequency. Thanks for accurately pin-pointing this while reviewing! I've changed my mind and give this some love. :)

mvines commented on ci/dependabot-pr.sh:
```diff
@@ -0,0 +1,36 @@
+
+ci/docker-run.sh $rust_nightly_docker_image ci/dependabot-updater.sh
+
+if [[ $(git status --short :**/Cargo.lock | wc -l) -gt 0 ]]; then
```
I'd prefer an early return here:
```suggestion
if [[ $(git status --short :**/Cargo.lock | wc -l) -eq 0 ]]; then
  echo --- ok
  exit 0
fi
```
so the bigger iif block can be unindented

ryoqun commented:
@mvines I think I've fixed all your suggestions! Also, I've updated my experiment pr: https://github.com/solana-labs/solana/pull/9508 Off course, both builds are passing. :)

ryoqun commented:
I've rebased all outstading dependabot's PRs on this after the merge.Now, all those prs seem to be working: #9446 #9607.

ryoqun commented:
I'll wait a week or so before I back-port this to all version branches (It should be almost no effort).

Also, I'll do quick-post to the upstream's dependabot/feedback#5 to share this for the rust+dependabot community.
